src/python/WMCore/MicroService/Unified/MSOutputTemplate.py

```diff
@@ -0,0 +1,228 @@
+    #    https://stackoverflow.com/questions/32728670/mutex-with-mongodb
+    #    ```
+    #    db.collection.update({done: false, taken: false},{$set: {taken: true, takenBy: myIp}});
+    #    ```
+
+    # TODO:
+    #    as soon as any change in the instantiated object happens ( any call of
+    #    a set method) the same change must be updated in Mongodb too -
+    #    this may happen from outside or from the object itself. The former method
+    #    keeps the document independent from the underlying Database technology
+
+    def __init__(self, doc=None, **kwargs):
+        super(MSOutputTemplate, self).__init__(**kwargs)
+        # super(MSOutputTemplate, self).__init__()
+
+        docTemplate = [
+            ('_id', None, unicode),
+            ('RequestName', None, unicode),
+            ('RequestStatus', None, unicode),
+            ('creationTime', None, Timestamp),
+            ('lastUpdate', None, float),
+            ('isRelVal', None, bool),
+            ('isTaken', None, bool),
+            ('OutputDatasets', None, list),
+            ('destination', None, list),
+            ('destinationOutputMap', None, dict),
+            ('campaignOutputMap', None, dict),
+            ('transferStatus', None, unicode),
+            ('transferIDs', None, unicode),
+            ('numberOfCopies', None, int)]
+
+        self.docTemplate = docTemplate
+        self.required = ['_id', 'RequestName', 'creationTime']
+        self.allowEmpty = ['OutputDatasets']
+
+        myDoc = dict()
+        for tup in docTemplate:
+            myDoc.update({tup[0]: tup[1]})
+
+        # set creation time - if already present in the document it will be
+        # overwritten with the value from the document itself during the _checkAttr
+        # call and this value here will be ignored
+        myDoc['creationTime'] = Timestamp(int(time.time()), 1)
+
+        # if no document was passed consider only **kwargs
+        if doc is not None:
+            # Search the keys from the document template for their equivalent into
+            # the document passed and fill them into internal document so that they
+            # can pass the needed checks later, throw the unneeded/unrecognised
+            # key/values from the passed document
+            for key in map(lambda tup: tup[0], docTemplate):
+                if key in doc.keys():
+                    myDoc[key] = deepcopy(doc[key])
+
+        # if any object attribute is passed as a **kwargs parameter then
+        # overwrite the equivalent parameter which was coming with the doc
+        self._checkAttr(docTemplate, myDoc, update=True, throw=True, **kwargs)
+
+        # enforce a full check on the final document
+        self._checkAttr(docTemplate, myDoc, update=False, throw=True, **myDoc)
+
+        # final validation:
+        if self._checkValid(myDoc, throw=True):
+            self.update(myDoc)
+
+    def checkAttr(self):
```

Comment:
is this method ever called?

Reply:
It is not called anywhere in the code, but I put it here because while developing and testing it turned out to be quite useful. Especially when I mess up with the documents directly in iPython and try to break stuff intentionally, at some point I need to be able to see what I am doing so I just call this one to provoke a self consistency check of the object. This works quite well.